### ISSUE TITLE
Unify `ScalarUDF` and `ScalarFunction`

### ISSUE THREAD

**Issue by jackwener** [apache/datafusion#6282]
### Is your feature request related to a problem or challenge?

from https://github.com/apache/arrow-datafusion/pull/6281#discussion_r1187572265 @alamb 

In fact, ScalarUDF and ScalarFunction both are function, and the usage is also similar.

Perhaps we might provide a way to unify them.

**Comment by alamb** [comment]
I don't think this issue is completed quite yet

**Comment by alamb** [comment]
In particular, I think unifying these implementations would make it easier to add both built in functions (like what @izveigor is doing in https://github.com/apache/arrow-datafusion/pull/6149) as well as user defined functions -- as it would ensure that user defined functions had the same level of expressiveness / functionality as built in ones.

**Comment by Jefffrey** [comment]
Are we good to close this now that `ScalarFunction` essentially wraps `ScalarUDF`?

https://github.com/apache/datafusion/blob/0ae9fdcb0da0cfaf180848115705fe3a9b7de343/datafusion/expr/src/expr.rs#L967-L976

**Comment by alamb** [comment]
Yes, I think so. I don't think we have any plans to cause code churn in this area in the near future so let's close this one
